src/System.Text.Json/tests/Serialization/DictionaryTests.cs

```diff
@@ -105,14 +106,53 @@ public static void DuplicateKeysFail()
         [Fact]
         public static void DictionaryOfObject()
         {
-            Dictionary<string, object> obj = JsonSerializer.Parse<Dictionary<string, object>>(@"{""Key1"":1}");
-            Assert.Equal(1, obj.Count);
-            JsonElement element = (JsonElement)obj["Key1"];
-            Assert.Equal(JsonValueType.Number, element.Type);
-            Assert.Equal(1, element.GetInt32());
+            {
+                Dictionary<string, object> obj = JsonSerializer.Parse<Dictionary<string, object>>(@"{""Key1"":1}");
+                Assert.Equal(1, obj.Count);
+                JsonElement element = (JsonElement)obj["Key1"];
+                Assert.Equal(JsonValueType.Number, element.Type);
+                Assert.Equal(1, element.GetInt32());
 
-            string json = JsonSerializer.ToString(obj);
-            Assert.Equal(@"{""Key1"":1}", json);
+                string json = JsonSerializer.ToString(obj);
+                Assert.Equal(@"{""Key1"":1}", json);
+            }
+
+            {
+                IDictionary<string, object> obj = JsonSerializer.Parse<IDictionary<string, object>>(@"{""Key1"":1}");
+                Assert.Equal(1, obj.Count);
+                JsonElement element = (JsonElement)obj["Key1"];
+                Assert.Equal(JsonValueType.Number, element.Type);
+                Assert.Equal(1, element.GetInt32());
+
+                string json = JsonSerializer.ToString(obj);
+                Assert.Equal(@"{""Key1"":1}", json);
+            }
+        }
+
+        [Fact]
+        public static void ImplementsIDictionaryOfObject()
+        {
+            var input = new StringToObjectIDictionaryWrapper(new Dictionary<string, object>
```

Comment:
Can you test also a parse...I think it will break [here](https://github.com/dotnet/corefx/pull/38512/files#diff-cf3c98b9ccd0b5614511b8adf262f8e9R65) , I did some tests and seem that we need to discover key/value with some reflection.

Reply:
Something like [this](https://github.com/dotnet/corefx/compare/master...MarcoRossignoli:json_dictionary?diff=unified#diff-cf3c98b9ccd0b5614511b8adf262f8e9R66) but with two path `IDictionaryEnumerator` where we can access to key,value and `IEnumerable` with some reflection or maybe emitting performant il

Reply:
try also this test https://github.com/dotnet/corefx/compare/master...MarcoRossignoli:json_dictionary?diff=unified#diff-7ee1539dfcc7bc8b662ab8bf8c3a91c1R654

Reply:
> Can you test also a parse...I think it will break here

[Parsing](https://github.com/dotnet/corefx/blob/e206be99ec8fe63ea6a4f9ea7f1e7bfd3dcb59c9/src/System.Text.Json/tests/Serialization/DictionaryTests.cs#L144-L147) works.

> try also this test https://github.com/dotnet/corefx/compare/master...MarcoRossignoli:json_dictionary?diff=unified#diff-7ee1539dfcc7bc8b662ab8bf8c3a91c1R654

This works also. Since the dict value is object, it ultimately does down [this code path](https://github.com/dotnet/corefx/blob/e206be99ec8fe63ea6a4f9ea7f1e7bfd3dcb59c9/src/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Write.HandleDictionary.cs#L55).

Reply:
Well done!I didn't touch AddProperty so I think I took another code path and get cast exception.